packages/optimizer/lib/transformers/MinifyHtml.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const Terser = require('terser');
+const {remove} = require('../NodeUtils');
+
+// Ignore comments of the form <!-- __AAAA_BBBB___ --> by default (used by Next.js)
+const COMMENT_DEFAULT_IGNORE = /^\s*__[a-bA-Z0-9_-]+__\s*$/;
+
+// Matches all consecutive whitesapce
+const WHITESPACE_REGEX = /\s+/g;
+
+/**
+ * MinifyHtml - minifies files size by:
+ *
+ * - minifying inline JSON
+ * - minifying inline amp-script using https://www.npmjs.com/package/terser
+ * - collapsing whitespace outside of pre, script, style and area.
+ * - removing comments
+ *
```

Comment:
Here's a list of additional transformations you could consider:
- removing quotes where they are not needed (`lang="en_US"` => `lang=en_US`)
- removing empty attribute values (`data-bool-flag=""` => `data-bool-flag`)
- removing unneeded attribute values (`disabled="disabled"` => `disabled`)
- removing redundant attribute values where the values is the default (`<input type=number step=1>` => `<input type=number>`)
- removing optional tags: https://html.spec.whatwg.org/multipage/syntax.html#optional-tags
- decoding HTML entities into Unicode characters (`&copy;` => `©`)

Probably not needed within an Amp context:
- removing `type="text/javascript"` from `<script>` tags
- removing `type="text/css"` from `<style>` tags

Reply:
Thanks for the suggestions! I've filed https://github.com/ampproject/amp-toolbox/issues/594

Some things are already handled by the parser (e.g. removing empty attribute values)